diff -Naur ppp-2.4.4.0/pppd/chap-new.c ppp-2.4.4/pppd/chap-new.c
--- ppp-2.4.4.0/pppd/chap-new.c	2012-01-30 20:52:16.389255467 +0800
+++ ppp-2.4.4/pppd/chap-new.c	2012-01-31 19:23:26.087948094 +0800
@@ -36,6 +36,8 @@
 #include "chap-new.h"
 #include "chap-md5.h"
 
+#include "syncppp.h"
+
 #ifdef CHAPMS
 #include "chap_ms.h"
 #define MDTYPE_ALL (MDTYPE_MICROSOFT_V2 | MDTYPE_MICROSOFT | MDTYPE_MD5)
@@ -58,6 +60,7 @@
 int chap_max_transmits = 10;
 int chap_rechallenge_time = 0;
 int chapms_strip_domain = 0;
+bool syncpppd = 0;
 
 /*
  * Command-line options.
@@ -71,6 +74,8 @@
 	  "Set interval for rechallenge", OPT_PRIO },
 	{ "chapms-strip-domain", o_bool, &chapms_strip_domain,
 	  "Strip the domain prefix before the Username", 1 },
+    { "syncppp", o_bool, &syncpppd,
+      "sync among multiple pppd when sending chap respond", 1 },
 	{ NULL }
 };
 
@@ -475,6 +480,14 @@
 	p[2] = len >> 8;
 	p[3] = len;
 
+    if (syncpppd) {
+        if (syncppp() < 0) {
+            error("syncppp sync fail");
+        } else {
+            info("syncppp sync succeeded");
+        }   
+    }
+
 	output(0, response, PPP_HDRLEN + len);
 }
 
diff -Naur ppp-2.4.4.0/pppd/Makefile.linux ppp-2.4.4/pppd/Makefile.linux
--- ppp-2.4.4.0/pppd/Makefile.linux	2012-01-30 20:52:16.389255467 +0800
+++ ppp-2.4.4/pppd/Makefile.linux	2012-01-30 20:47:24.714744804 +0800
@@ -13,16 +13,16 @@
 
 PPPDSRCS = main.c magic.c fsm.c lcp.c ipcp.c upap.c chap-new.c md5.c ccp.c \
 	   ecp.c ipxcp.c auth.c options.c sys-linux.c md4.c chap_ms.c \
-	   demand.c utils.c tty.c eap.c chap-md5.c
+	   demand.c utils.c tty.c eap.c chap-md5.c syncppp.c
 
 HEADERS = ccp.h chap-new.h ecp.h fsm.h ipcp.h \
 	ipxcp.h lcp.h magic.h md5.h patchlevel.h pathnames.h pppd.h \
-	upap.h eap.h
+	upap.h eap.h syncppp.h
 
 MANPAGES = pppd.8
 PPPDOBJS = main.o magic.o fsm.o lcp.o ipcp.o upap.o chap-new.o md5.o ccp.o \
 	   ecp.o auth.o options.o demand.o utils.o sys-linux.o ipxcp.o tty.o \
-	   eap.o chap-md5.o
+	   eap.o chap-md5.o syncppp.o
 
 #
 # include dependencies if present
@@ -33,7 +33,7 @@
 # CC = gcc
 #
 COPTS = -O2 -pipe -Wall -g
-LIBS =
+LIBS = -lpthread
 
 # Uncomment the next 2 lines to include support for Microsoft's
 # MS-CHAP authentication protocol.  Also, edit plugins/radius/Makefile.linux.
@@ -70,6 +70,9 @@
 # Enable plugins
 PLUGIN=y
 
+# Enable SYNC_PPP plugins
+SYNC_PPP=y
+
 # Enable Microsoft proprietary Callback Control Protocol
 #CBCP=y
 
@@ -97,6 +100,13 @@
 endif
 endif
 
+# syncppp 
+ifdef SYNC_PPP
+TARGETS	+= syncpppinit
+EXTRAINSTALL = $(INSTALL) -c -m 555 syncpppinit $(BINDIR)/syncpppinit
+EXTRACLEAN += syncpppinit.o
+endif
+
 # EAP SRP-SHA1
 ifdef USE_SRP
 CFLAGS	+= -DUSE_SRP -DOPENSSL -I/usr/local/ssl/include
@@ -219,6 +229,9 @@
 pppd: $(PPPDOBJS)
 	$(CC) $(CFLAGS) $(LDFLAGS) -o pppd $(PPPDOBJS) $(LIBS)
 
+syncpppinit:	syncpppinit.c
+	${CC} $(CFLAGS) $(LDFLAGS) -o syncpppinit syncpppinit.c $(LIBS)
+
 srp-entry:	srp-entry.c
 	$(CC) $(CFLAGS) $(LDFLAGS) -o $@ srp-entry.c $(LIBS)
 
diff -Naur ppp-2.4.4.0/pppd/syncppp.c ppp-2.4.4/pppd/syncppp.c
--- ppp-2.4.4.0/pppd/syncppp.c	1970-01-01 08:00:00.000000000 +0800
+++ ppp-2.4.4/pppd/syncppp.c	2012-01-31 19:17:42.579302502 +0800
@@ -0,0 +1,56 @@
+#include <stdio.h>
+#include <sys/types.h>
+#include <sys/shm.h>
+#include <sys/ipc.h>
+#include <stdlib.h>
+#include <semaphore.h>
+#include <fcntl.h>
+#include <sys/file.h>
+#include "pppd.h"
+#include "syncppp.h"
+#include <sys/stat.h>
+#include <unistd.h>
+
+int syncppp(void)
+{
+    int shm_id;
+    key_t key;
+    int fdlock;
+    struct semaphores *semphs;
+
+    if ((key = ftok(keyfilename, PROJ_ID)) == -1) {
+        error("ftok key error");
+        return -1;
+    }
+
+    if ((shm_id = shmget(key, 1, 0644)) < 0) {
+        error("shmget error");
+        return -1;
+    }
+
+    if ( (void *)(semphs = shmat(shm_id, 0, 0)) == (void *)-1) {
+        error("shmat error");
+        return -1;
+    }
+
+    if ((sem_post(&(semphs->count))) < 0) {
+        error("sem_post error");
+        return -1;
+    }
+
+    shmdt(semphs);
+
+    if ((fdlock = open(lockfilename,O_RDONLY, 0644)) < 0) {
+        error("lockfile open error");
+        return -1;
+    }
+
+    if (flock(fdlock,LOCK_SH) < 0) {
+        error("flock error");
+        return -1;
+    }
+    close(fdlock);
+
+    return 0;
+
+}
diff -Naur ppp-2.4.4.0/pppd/syncppp.h ppp-2.4.4/pppd/syncppp.h
--- ppp-2.4.4.0/pppd/syncppp.h	1970-01-01 08:00:00.000000000 +0800
+++ ppp-2.4.4/pppd/syncppp.h	2012-01-31 19:18:04.829646549 +0800
@@ -0,0 +1,13 @@
+#include<semaphore.h>
+
+#define MAX_PPP_NUM 30
+#define PROJ_ID 225
+#define keyfilename "/tmp/pppkeyfile"
+#define lockfilename "/tmp/ppplockfile"
+
+int syncppp(void);
+
+struct semaphores {
+    sem_t count;  /* count the pppd processes which has recieved the challenge */
+};
+
diff -Naur ppp-2.4.4.0/pppd/syncpppinit.c ppp-2.4.4/pppd/syncpppinit.c
--- ppp-2.4.4.0/pppd/syncpppinit.c	1970-01-01 08:00:00.000000000 +0800
+++ ppp-2.4.4/pppd/syncpppinit.c	2012-01-31 19:17:53.396136452 +0800
@@ -0,0 +1,97 @@
+#include<stdio.h>
+#include<sys/types.h>
+#include<sys/shm.h>
+#include<sys/ipc.h>
+#include<stdlib.h>
+#include <semaphore.h>
+#include <fcntl.h>
+#include <sys/file.h>
+#include <sys/stat.h>
+#include<unistd.h>
+#include "syncppp.h"
+
+int openlockfile()
+{
+    int fdlock;
+
+    if ((fdlock = creat(lockfilename, 0644)) < 0) {
+        perror("fdlock open error");
+        exit(1);
+    }
+    return fdlock;
+}
+
+void lockfile(int fdlock)
+{
+    if (flock(fdlock, LOCK_EX) < 0) {
+        perror("flock lock error");
+        exit(1);
+    }    
+    fprintf(stderr,"syncpppinit: locked\n");
+}
+
+void unlockfile(int fdlock)
+{
+    if (flock(fdlock, LOCK_UN) < 0) {
+        perror("flock unlock error");
+        exit(1);
+    }
+}
+
+
+int main(int argc, char *argv[])
+{
+    int shm_id;
+    int ppp_num;
+    int fdlock;
+    sem_t *p_sem;
+    key_t key;
+    struct semaphores *semphs;
+
+    if (argc != 2) {
+        fprintf(stderr, "Usage: %s <number of pppd>\n", argv[0]);
+        exit(1);
+    }
+    
+    ppp_num = atoi(argv[1]);
+    if (ppp_num > MAX_PPP_NUM || ppp_num <= 0) {
+        fprintf(stderr, "Number of pppd beyoung limit\n");
+        exit(1);
+    }
+
+    /* create a uniqe key */
+    creat(keyfilename, 0755);
+    if ((key = ftok(keyfilename, PROJ_ID)) == -1) {
+        perror("key error");
+        exit(1);
+    }
+
+    shm_id = shmget(key, sizeof(struct semaphores), IPC_CREAT | IPC_EXCL | 0644);
+    if (shm_id < 0) {
+        /* exist */
+        shm_id = shmget(key, 1, 0644);
+    }
+
+    if ( (void *)(semphs = shmat(shm_id, 0, 0)) == (void *)-1) {
+        perror("shmat error");
+        exit(1);
+    }
+
+    if (sem_init(&(semphs->count), 1, 0) < 0) {
+        perror("sem_init error"); /* shared between processes, init 0 */
+        exit(1);
+    }
+
+    fdlock = openlockfile();
+    lockfile(fdlock);
+
+    while (ppp_num > 0) {
+        sem_wait(&(semphs->count));
+        fprintf(stderr,"%d ",ppp_num);
+        ppp_num--;
+    }
+    unlockfile(fdlock);
+    fprintf(stderr,"\nsyncpppinit: unlocked\n");
+
+    return 0;
+}
